tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/FoldOperLetinizer.scala

```diff
@@ -0,0 +1,64 @@
+package at.forsyte.apalache.tla.pp
+
+import at.forsyte.apalache.tla.lir.{LetInEx, NameEx, OperEx, TlaEx, TlaOperDecl}
+import at.forsyte.apalache.tla.lir.oper.ApalacheOper
+import at.forsyte.apalache.tla.lir.storage.BodyMap
+import at.forsyte.apalache.tla.lir.transformations.standard.DeepCopy
+import at.forsyte.apalache.tla.lir.transformations.{TlaExTransformation, TransformationTracker}
+
+/**
+ * Replaces FoldSet( A, ..., ...), where A(p,q) == X with FoldSet( LET localA(p,q) == X IN localA, ..., ...)
+ *
+ * The purpose of this is to embed the operator definition at the call-by-name site, to allow for
+ * a) the deletion of A from the operator declaration set
+ * b) the removal of a BodyMap dependency in FoldSet rewriting.
+ *
+ * <p>Example: `A(p,q) == p + q`, `FoldSet(A, 0, {1,2,3})` becomes
+ * `FoldSet( LET localA(p,q) == p + q IN localA, 0, {1,2,3})`.
+ * </p>
+ *
+ * @author Jure Kukovec
+ */
+class FoldOperLetinizer(tracker: TransformationTracker, operMap: BodyMap, nameGenerator: UniqueNameGenerator)
```

Comment:
Shall we think about generalization of this class?

Reply:
This is basically what I plan to do to resolve #872.